binr/r2r/load.c

```diff
@@ -0,0 +1,160 @@
+/* radare - LGPL - Copyright 2020 - thestr4ng3r */
+
+#include "r2r.h"
+
+#define LINEFMT "%s, line %"PFMT64u": "
+
+R_API R2RCmdTest *r2r_cmd_test_new() {
+	return R_NEW0 (R2RCmdTest);
+}
+
+R_API void r2r_cmd_test_free(R2RCmdTest *test) {
+	if (!test) {
+		return;
+	}
+#define DO_KEY_STR(key, field) free (test->field.value);
+	R2R_CMD_TEST_FOREACH_RECORD(DO_KEY_STR, R2R_CMD_TEST_FOREACH_RECORD_NOP)
+#undef DO_KEY_STR
+	free (test);
+}
+
+// read the (possibly multiline) string value of some key in the file
+// e.g. for
+//
+// 0    CMDS=<<EOF
+// 1    Hello
+// 2    World
+// 3    EOF
+// 4    ...
+//
+// if f is at the beginning of line 1,
+// read_string_val(f, "<<EOF\0")
+// will return "Hello\nWorld\n" with f being at the beginning of line 4 afterwards.
+static char *read_string_val(FILE *f, const char *val, ut64 *linenum) {
+	if (val[0] == '\'') {
+		eprintf ("Error: Invalid string syntax, use <<EOF instead of '...'\n");
+		return NULL;
+	}
+	if (val[0] == '<' && val[1] == '<') {
+		// <<EOF syntax
+		const char *endtoken = val + 2;
+		if (!*endtoken) {
+			eprintf ("Error: Missing opening end token after <<\n");
+			return NULL;
+		}
+		RStrBuf *buf = r_strbuf_new ("");
+		char *line = NULL;
+		size_t linesz = 0;
+		ssize_t len;
+		while ((len = getline (&line, &linesz, f)) >= 0) {
+			(*linenum)++;
+			char *end = strstr (line, endtoken);
```

Comment:
end token can be only EOF. we enforced this for simplicity. if the token is not  EOF you may throw a warning like i did  in r2r.v

Reply:
What if you have a multiline string that contains EOF?